explainers.md

```diff
@@ -0,0 +1,860 @@
+
+Also, if manifest\_v1 is installed on an offline device 1 and manifest\_v2 is installed on an offline device 2, and sync is later turned on, it will result in temporary duplicated apps before the browser finishes the manifest update and dedupe it.  In general it’s very error prone to make the primary key changeable.
+
+Overall, if we use manifest\_url as the global\_id, we need to **not support **the use case of making manifest\_url updatable or deal with **non stable primary keys** for managing apps across devices. The upside of this approach is that the manifest\_url is a** resolvable URL**. The global\_id can be used alone to discover the app, however that’s a “nice to have” feature, but not a core requirement for the global\_id. Another benefit is that it allows non user agent entities having a simpler flow to keep the app up to date - they just need to save the manifest\_url as the key and periodically fetch it. 
+
+example.com global\_id:
+
+
+```
+https://www.example.com/manifest.webmanifest
+```
+
+
+example.com with cdn-manifest global\_id:
+
+
+```
+https://www.real-cdn.com/example-com-manifest.webmanifest
+```
+
+
+Below pros and cons use the first option to support 301 redirects - that is, always use the original manifest\_url as global\_id. As the second approach is not reliable.
+
+Pros
+
+*   Simplicity. A single existing field can be used instead of introducing a new field.
+*   The id is a meaningful URL, external parties that use this id to reference the web app can use it directly to get the metadata of the web app.
+*   Allows [S5: apps to change start\_url](#scenario-5) without special handling.
+
+Cons	
+
+*   [S2: installed apps on desktop](#scenario-2), [S3: Apps that are installed on both Android and desktop](#scenario-3) needs migration.
+*   Change of implicit expectations for desktop WebApp developers
+*   **[S6: Apps that specify manifest as data URL](#scenario-6) will not be updatable**
+*   PWA stores will recognize different versions of manifest URLs as different web apps. Alternatively they can load the start\_url to find the manifest link, and follow redirect to find the latest manifest\_url and use that as the authoritative manifest. 
+*   [S4: Apps that need to update manifest URLs](#scenario-4) will need to keep supporting the previous manfiest\_url, handle 301 to the latest manifest forever. The document will need to always use the original manifest\_url. An extra roundtrip will always be needed for loading the app.
+*   If the manifest\_url is hosted under a CDN, they will need to stick to that CDN provider.
+*   Difficult to reconcile with [S8: apps installed from A2HS](#scenario-8), which does NOT require a manifest. ID creating here is weird, and instead would have to have custom upgrade handling for when a manifest is found on the page.
+
+
```

Comment:
Might be nice to add a 

Note: If there weren't existing apps & infrastructure that relied on a changing manifest_url, and this was being designed from the beginning, this option is a lot more attractive.

Reply:
added